src/main/java/com/waytta/SaltAPIBuilder.java

```diff
@@ -47,515 +44,561 @@
+                }
+                returnArray = httpResponse.getJSONArray("return");
+                // Check the return[] array to see how many minions have
+                // responded
+                if (!returnArray.getJSONObject(0).names().isEmpty()) {
+                    numMinionsDone = returnArray.getJSONObject(0).names().size();
+                } else {
+                    numMinionsDone = 0;
+                }
+                listener.getLogger().println(numMinionsDone + " minions are done");
+            } catch (Exception e) {
+                listener.getLogger()
+                        .println("Problem: " + myfunction + " " + myarguments + " to " + myservername
+                                + " for " + mytarget + ":\n" + e + "\n\n" + httpResponse.toString(2));
+                return false;
+            }
+
+            // Now that we know how many minions have responded, and how many we
+            // are waiting on. Let's see more have finished
+            if (numMinionsDone < numMinions) {
+                // Don't print annying messages unless we really are waiting for
+                // more minions to return
+                listener.getLogger().println(
+                        "Will check status every " + String.valueOf(myJobPollTime) + " seconds...");
+            }
+            while (numMinionsDone < numMinions) {
+                try {
+                    Thread.sleep(myJobPollTime * 1000);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                    // Allow user to cancel job in jenkins interface
+                    listener.getLogger().println("Cancelling job");
+                    return false;
+                }
+                httpResponse = Utils.getJSON(myservername + "/jobs/" + jid, null, token);
+                try {
+                    returnArray = httpResponse.getJSONArray("return");
+                    numMinionsDone = returnArray.getJSONObject(0).names().size();
+                } catch (Exception e) {
+                    listener.getLogger()
+                            .println("Problem: " + myfunction + " " + myarguments + " for " + mytarget
+                                    + ":\n" + e + "\n\n" + httpResponse.toString(2).split("\\\\n")[0]);
+                    return false;
+                }
+            }
+            if (returnArray.get(0).toString().contains("TypeError")) {
+                listener.getLogger().println("Salt reported an error for " + myfunction + " "
+                        + myarguments + " for " + mytarget + ":\n" + returnArray.toString(2));
+                return false;
+            }
+
+            // If running script module check for non-zero exit codes in
+            // retcode.
+            returnArray = httpResponse.getJSONArray("return");
+            boolean validFunctionExecution = Utils.validateFunctionCall(returnArray);
```

Comment:
I'm not in front of a salt environment at the moment. Has this bit been tested, that all functions set a proper return code? 
test.ping
test.arg
cmd.run
cmd.script
state.sls
Plus the salt runner for state.orchestrate